usb: Fix ACL data tx stalling

mkarhumaa opened the pull request:
This patch increases size of acl_tx_pool and makes sure that acl_read_cb
waits for available buffers.

Previously it was possible to run out of buffers and in that case
acl_read_cb returned. This caused ACL data TX to stall because device
did not read data from HCI out endpoint.

Fixes #14899

Signed-off-by: Matias Karhumaa <matias.karhumaa@gmail.com>

mkarhumaa committed:
usb: Fix ACL data tx stalling

This patch increases size of acl_tx_pool and makes sure that acl_read_cb
waits for available buffers.

Size of acl_tx_pool is now configurable with CONFIG_BT_CTLR_TX_BUFFERS.

Previously it was possible to run out of buffers and in that case
acl_read_cb returned. This caused ACL data TX to stall because device
did not read data from HCI out endpoint.

Fixes #14899

Signed-off-by: Matias Karhumaa <matias.karhumaa@gmail.com>

jhedberg commented on subsys/usb/class/bluetooth.c:
```diff
@@ -38,7 +38,7 @@ NET_BUF_POOL_DEFINE(tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
 /* Data size needed for ACL buffers */
 #define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
-NET_BUF_POOL_DEFINE(acl_tx_pool, 2, BT_BUF_ACL_SIZE, sizeof(u8_t), NULL);
+NET_BUF_POOL_DEFINE(acl_tx_pool, 4, BT_BUF_ACL_SIZE, sizeof(u8_t), NULL);
```
Would it make sense to sync this up with `CONFIG_BT_CTLR_TX_BUFFERS`?

mkarhumaa replied:
Ok, sounds feasible. I will fix it.

cvinayak replied:
is increase to 4 or to CONFIG_BT_CTLR_TX_BUFFERS justified?  implicit flow control by use of K_FOREVER should be enough to throttle ACL packet into controller, hence the original double buffer should suffice?

jhedberg replied:
@mkarhumaa note that `CONFIG_BT_CTLR_TX_BUFFERS` will not be available in some "exotic" configurations, i.e. when you're using USB to proxy a controller that's behind some physical transport rather than the native controller. So you'll probably want hci_usb to have it's own define something along the lines of:

```
#if defined(CONFIG_BT_CTLR_TX_BUFFERS)
#define ACL_BUF_COUNT CONFIG_BT_CTLR_TX_BUFFERS
#else
#define ACL_BUF_COUNT 4
#endif
```

jhedberg replied:
> is increase to 4 or to CONFIG_BT_CTLR_TX_BUFFERS justified? 

@cvinayak that's a fair question. I was just thinking that if one sets `CONFIG_BT_CTLR_TX_BUFFERS` to a considerably higher value, then hci_usb will become a bottleneck if its own pool doesn't grow dynamically.

mkarhumaa replied:
> is increase to 4 or to CONFIG_BT_CTLR_TX_BUFFERS justified? implicit flow control by use of K_FOREVER should be enough to throttle ACL packet into controller, hence the original double buffer should suffice?

This was suggested by Johann F. at #usb channel in Slack. I think increasing acl_tx_pool should make sure that it will not become bottleneck and stall on host side.

cvinayak replied:
ok, its the speed of USB+DMA vs the memcpy into controller's tx buffers. I would use a low buffer count, and users can be instructed in the form of comment to increase the buffer count upto the CTLR_TX_BUFFERS if degraded throughput is observed. (lets try to keep low RAM usage in our samples).

jhedberg replied:
@cvinayak isn't also USB+DMA vs how fast the packets are sent over air and the buffers consequently freed up in the controller? I'd imagine the memcpy to be negligible there.

mkarhumaa replied:
@cvinayak So you prefer keeping acl_tx_pool size in 2 and just adding comment? To my understanding CONFIG_BT_CTLS_RX_BUFFERS defaults to 7 if BT_HCI_RAW is enabled and 3 if not.

Keeping it in 2 is fine for me.

cvinayak commented on subsys/usb/class/bluetooth.c:
```diff
@@ -188,7 +188,7 @@ static void acl_read_cb(u8_t ep, int size, void *priv)
 	}
 
-	buf = net_buf_alloc(&acl_tx_pool, K_NO_WAIT);
+	buf = net_buf_alloc(&acl_tx_pool, K_FOREVER);
```
instead of logging error, use __ASSERT_NO_MSG, that is how it is done in bluetooth subsystem.

mkarhumaa replied:
Will fix this.

jhedberg replied:
Yep. Assert should be used for things which indicate a bug somewhere, and getting NULL when you have K_FOREVER is definitely a bug.

finikorg reviewed (approved):
LGTM

mkarhumaa commented:
I have force-pushed a version addressing the comments. I ended up using `CONFIG_BT_CTLR_TX_BUFFERS` to configure size of `acl_tx_pool`. Let me know if there is anything else that should be fixed.

cvinayak reviewed (approved):
LGTM
